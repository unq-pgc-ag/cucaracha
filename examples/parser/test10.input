fun f() : Int {
  return 1
}

fun g(x : Int) : Int {
  return x
}

fun h(x : Int, y : Int) : Int {
  return x + y
}

fun i(x : Int, y : Int, z : Int) : Int {
  return x * y * z
}

fun beta(x:Bool, y:Bool) : Int {
  if x and y {
    z := 1
  } else {
    z := 0
  }
  return z
}

fun main() {
  v := [1,2,3]
  x := f()
  x := g(100)
  x := g(100 + 100)
  x := g(v[0])
  x := g(#v)
  x := g(#v + 100)
  x := g(f())
  x := h(11,12)
  x := h(v[0],v[1])
  x := h(f(),f())
  x := h(f(),g(f()))
  x := h(f(),g(#v))
  x := h(f(),v[f()])
  x := h(g(f()),f())
  x := h(g(#v),f())
  x := h(v[f()],f())
  x := h(g(f()),g(f()))
  x := h(g(v[f()]),v[f()])
  x := h(v[f()],v[f()])
  x := v[h(v[f()],v[f()])]
  x := i(1,2,3)
  x := v[i(1,2,3)]
  x := v[i(#v,#v + 1,#v+2)]
  x := i(h(1,11),h(2,22),h(3,33))
  x := v[i(h(1,11),h(2,22),h(3,33))]
  x := v[i(h(1,11),h(2,22),h(3,33))]
  x := v[i(h(1,11),h(2,22),h(3,33))]*#v
  x := beta(1 != 2 or 2 != 1, 1 < 2 or 1 > 2)
}

