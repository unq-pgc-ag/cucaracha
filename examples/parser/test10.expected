(Program
  (Function
    f
    Int
    (Block
      (StmtReturn
        (ExprConstNum
          1
        )
      )
    )
  )
  (Function
    g
    Int
    (Parameter
      x
      Int
    )
    (Block
      (StmtReturn
        (ExprVar
          x
        )
      )
    )
  )
  (Function
    h
    Int
    (Parameter
      x
      Int
    )
    (Parameter
      y
      Int
    )
    (Block
      (StmtReturn
        (ExprAdd
          (ExprVar
            x
          )
          (ExprVar
            y
          )
        )
      )
    )
  )
  (Function
    i
    Int
    (Parameter
      x
      Int
    )
    (Parameter
      y
      Int
    )
    (Parameter
      z
      Int
    )
    (Block
      (StmtReturn
        (ExprMul
          (ExprMul
            (ExprVar
              x
            )
            (ExprVar
              y
            )
          )
          (ExprVar
            z
          )
        )
      )
    )
  )
  (Function
    beta
    Int
    (Parameter
      x
      Bool
    )
    (Parameter
      y
      Bool
    )
    (Block
      (StmtIfElse
        (ExprAnd
          (ExprVar
            x
          )
          (ExprVar
            y
          )
        )
        (Block
          (StmtAssign
            z
            (ExprConstNum
              1
            )
          )
        )
        (Block
          (StmtAssign
            z
            (ExprConstNum
              0
            )
          )
        )
      )
      (StmtReturn
        (ExprVar
          z
        )
      )
    )
  )
  (Function
    main
    Unit
    (Block
      (StmtAssign
        v
        (ExprVecMake
          (ExprConstNum
            1
          )
          (ExprConstNum
            2
          )
          (ExprConstNum
            3
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          f
        )
      )
      (StmtAssign
        x
        (ExprCall
          g
          (ExprConstNum
            100
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          g
          (ExprAdd
            (ExprConstNum
              100
            )
            (ExprConstNum
              100
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          g
          (ExprVecDeref
            v
            (ExprConstNum
              0
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          g
          (ExprVecLength
            v
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          g
          (ExprAdd
            (ExprVecLength
              v
            )
            (ExprConstNum
              100
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          g
          (ExprCall
            f
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprConstNum
            11
          )
          (ExprConstNum
            12
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprVecDeref
            v
            (ExprConstNum
              0
            )
          )
          (ExprVecDeref
            v
            (ExprConstNum
              1
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            f
          )
          (ExprCall
            f
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            f
          )
          (ExprCall
            g
            (ExprCall
              f
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            f
          )
          (ExprCall
            g
            (ExprVecLength
              v
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            f
          )
          (ExprVecDeref
            v
            (ExprCall
              f
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            g
            (ExprCall
              f
            )
          )
          (ExprCall
            f
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            g
            (ExprVecLength
              v
            )
          )
          (ExprCall
            f
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprVecDeref
            v
            (ExprCall
              f
            )
          )
          (ExprCall
            f
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            g
            (ExprCall
              f
            )
          )
          (ExprCall
            g
            (ExprCall
              f
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprCall
            g
            (ExprVecDeref
              v
              (ExprCall
                f
              )
            )
          )
          (ExprVecDeref
            v
            (ExprCall
              f
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          h
          (ExprVecDeref
            v
            (ExprCall
              f
            )
          )
          (ExprVecDeref
            v
            (ExprCall
              f
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprVecDeref
          v
          (ExprCall
            h
            (ExprVecDeref
              v
              (ExprCall
                f
              )
            )
            (ExprVecDeref
              v
              (ExprCall
                f
              )
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          i
          (ExprConstNum
            1
          )
          (ExprConstNum
            2
          )
          (ExprConstNum
            3
          )
        )
      )
      (StmtAssign
        x
        (ExprVecDeref
          v
          (ExprCall
            i
            (ExprConstNum
              1
            )
            (ExprConstNum
              2
            )
            (ExprConstNum
              3
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprVecDeref
          v
          (ExprCall
            i
            (ExprVecLength
              v
            )
            (ExprAdd
              (ExprVecLength
                v
              )
              (ExprConstNum
                1
              )
            )
            (ExprAdd
              (ExprVecLength
                v
              )
              (ExprConstNum
                2
              )
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          i
          (ExprCall
            h
            (ExprConstNum
              1
            )
            (ExprConstNum
              11
            )
          )
          (ExprCall
            h
            (ExprConstNum
              2
            )
            (ExprConstNum
              22
            )
          )
          (ExprCall
            h
            (ExprConstNum
              3
            )
            (ExprConstNum
              33
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprVecDeref
          v
          (ExprCall
            i
            (ExprCall
              h
              (ExprConstNum
                1
              )
              (ExprConstNum
                11
              )
            )
            (ExprCall
              h
              (ExprConstNum
                2
              )
              (ExprConstNum
                22
              )
            )
            (ExprCall
              h
              (ExprConstNum
                3
              )
              (ExprConstNum
                33
              )
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprVecDeref
          v
          (ExprCall
            i
            (ExprCall
              h
              (ExprConstNum
                1
              )
              (ExprConstNum
                11
              )
            )
            (ExprCall
              h
              (ExprConstNum
                2
              )
              (ExprConstNum
                22
              )
            )
            (ExprCall
              h
              (ExprConstNum
                3
              )
              (ExprConstNum
                33
              )
            )
          )
        )
      )
      (StmtAssign
        x
        (ExprMul
          (ExprVecDeref
            v
            (ExprCall
              i
              (ExprCall
                h
                (ExprConstNum
                  1
                )
                (ExprConstNum
                  11
                )
              )
              (ExprCall
                h
                (ExprConstNum
                  2
                )
                (ExprConstNum
                  22
                )
              )
              (ExprCall
                h
                (ExprConstNum
                  3
                )
                (ExprConstNum
                  33
                )
              )
            )
          )
          (ExprVecLength
            v
          )
        )
      )
      (StmtAssign
        x
        (ExprCall
          beta
          (ExprOr
            (ExprNe
              (ExprConstNum
                1
              )
              (ExprConstNum
                2
              )
            )
            (ExprNe
              (ExprConstNum
                2
              )
              (ExprConstNum
                1
              )
            )
          )
          (ExprOr
            (ExprLt
              (ExprConstNum
                1
              )
              (ExprConstNum
                2
              )
            )
            (ExprGt
              (ExprConstNum
                1
              )
              (ExprConstNum
                2
              )
            )
          )
        )
      )
    )
  )
)
